content/courses/advanced-ada/chapters/limited_types.rst

```diff
@@ -6,9 +6,424 @@ Limited Types
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Limited_Types.Equality
+    :class: ada-expect-compile-error
+
+    package Limited_Types is
+
+       type Integer_Access is access Integer;
+
+       type Simple_Rec is limited private;
+
+       function Init (I : Integer) return Simple_Rec;
+
+    private
+
+       type Simple_Rec is limited record
+          V : Integer_Access;
+       end record;
+
+    end Limited_Types;
+
+    package body Limited_Types is
+
+       function Init (I : Integer) return Simple_Rec is
+       begin
+          return E : Simple_Rec do
+             E.V := new Integer'(I);
+          end return;
+       end Init;
+
+    end Limited_Types;
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Limited_Types; use Limited_Types;
+
+    procedure Show_Limited_Equality is
+       A : Simple_Rec := Init (5);
+       B : Simple_Rec := Init (6);
+    begin
+       if A = B then
+          Put_Line ("A = B");
+       else
+          Put_Line ("A /= B");
+       end if;
+    end Show_Limited_Equality;
+
+As expected, the comparison :ada:`A = B` triggers a compilation error because
+no predefined :ada:`=` operator is available for the :ada:`Simple_Rec` type.
+If we want to be able to compare objects of this type, we have to implement
+the :ada:`=` operator ourselves. For example, we can do that for the
+:ada:`Simple_Rec` type:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Limited_Types.Equality
+
+    package Limited_Types is
+
+       type Integer_Access is access Integer;
+
+       type Simple_Rec is limited private;
+
+       function Init (I : Integer) return Simple_Rec;
+
+       function "=" (Left, Right : Simple_Rec)
+                     return Boolean;
+
+    private
+
+       type Simple_Rec is limited record
+          V : Integer_Access;
+       end record;
+
+    end Limited_Types;
+
+    package body Limited_Types is
+
+       function Init (I : Integer) return Simple_Rec is
+       begin
+          return E : Simple_Rec do
+             E.V := new Integer'(I);
+          end return;
+       end Init;
+
+       function "=" (Left, Right : Simple_Rec)
+                     return Boolean is
+       begin
+          --  Comparing record components
+          return Left.V.all = Right.V.all;
+       end;
+
+    end Limited_Types;
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Limited_Types; use Limited_Types;
+
+    procedure Show_Limited_Equality is
+       A : Simple_Rec := Init (5);
+       B : Simple_Rec := Init (6);
+    begin
+       if A = B then
+          Put_Line ("A = B");
+       else
+          Put_Line ("A /= B");
+       end if;
+    end Show_Limited_Equality;
+
+Here, the :ada:`=` operator compares the dereferenced values of :ada:`Left.V`
+and :ada:`Right.V`, which matches our expectation for the :ada:`Simple_Rec`
+type. Declaring types limited ensures that only reasonable comparisons between
+objects of those types can be made.
```

Comment:
That's assuming the user-defined equality is reasonable, which isn't necessarily so. So perhaps it would be good to say: "Declaring types as limited ensures that unreasonable equality comparisons, and allows us to create reasonable replacements when required."

Reply:
Yes, thanks: I adapted the text accordingly.